### DISCUSSION TITLE
Will there be an open source introductory book about mojo somewhere in the future?

### DISCUSSION THREAD

**Discussion by slf188** [modular/modular#85] [Ideas]
Yesterday I created an issue and @goldiegadde recommended I post this as a new topic in the discussions section, so here it goes!

## Request
Mojo is still in the development stage but I wonder if it would be nice to boost the creation of an open-source introductory book. As someone interested in systems programming, I found it incredibly helpful that Rust has its own [open-source introductory book](https://doc.rust-lang.org/book/title-page.html) which back in the day helped me understand what Rust is capable of.

## Motivation
With an introductory book, imho more people would be able to learn and use Mojo, thus increasing its adoption and community support.

## Description and Requirements
Mojo integrates amazing new features, we can probably highlight those in the book and perhaps integrate practical examples and exercises to help readers apply what they've learned with an integrated editor the way [Will Crichton improved the existing Rust book](https://rust-book.cs.brown.edu/).

Overall, I think it would be a valuable resource for users, especially beginners to have this...why not? :)

I read the documentation available for Mojo and I think it's quite well documented, we should add something different to the book!

## Takeaways from Chris

@lattner was kind enough to bring up a good title "I'd particularly love to see a **Mojo for Python programmers** sort of document".

**Comment by Feanix-Fyre** [comment]
I spun up this repo today and I plan on working at it every day. I have some programming knowledge but no formal training (probably not the only one). One thing I've always found extremely beneficial in the process of teaching myself is just some general guide of how to progress. Is this that guide? I don't know. But being able to sort through communication issues with sLLMs (Silicon Large Language Model) has gotten me this far. I wonder what'll happen when I run it through a set of the best bLLMs (Biological LLMs) the universe is aware of.

And I don't really understand licenses but I don't want anybody, including myself, to create an Amazon e-book out of this. Not that I can stop someone from copy-pasting but, based on what I read, I think the license I chose protects the information in a way that prevents monetization.

As you will see once you get there though, that I'm totally winging it and so any suggestions or assistance would be nice. I'm trying to approach it from the perspective that someone accidentally clicked on a youtube video about mojo and they've never coded before in their life. Given that it's emerging tech, I felt like getting started on some type of curriculum and ensuring that it is public and easily accessible is the best course of action, and the best way that I can contribute at the moment.

So, yeah. *insert flame emoji* (How am I supposed to use this as a file extension if I don't know how to type it.....)
>Google "how to type flame emoji"
> %
yeah that alt code didn't work....

https://github.com/Feanix-Fyre/LearnMojoByFeanix

**Reply by TZDak** [reply]
> So, yeah. _insert flame emoji_ (How am I supposed to use this as a file extension if I don't know how to type it.....)

On Windows (at least 10 and 11) you can hold the Windows key and press period (AKA full stop) to get a menu into which you can type "fire" and get a bunch of emoji characters to pick from, or type "flame" and get ❤️‍🔥 and 🔥 to pick from. Generally though, I just look it up and then copy and paste, because that works with more than just emoji characters and I tend not to stop and categorize what kind of character that's not supported by my keyboard I'm looking for since my keyboard supports so few... and I'm using a multilingual keyboard.

DAK

**Reply by Feanix-Fyre** [reply]
woah. mind blown. thanks for that one.

**Reply by tristanbiesecker** [reply]
On Macs you can type `Command+Control+Spacebar` to get the emoji dialogue to pop up.

**Comment by lattner** [comment]
amazing, I'm so excited to see this effort

**Comment by tristanbiesecker** [comment]
I think this is a great idea and would love to help if hands are needed!

One completely personal request though, can we make it look a little more inviting and fresh (aesthetically) than the Rust site? The Rust book/site has always felt a bit old to me.

**Reply by slf188** [reply]
I completely agree with that! The Modular Website looks awesome! Let's make it align with how the modular website looks

**Reply by mojodojodev** [reply]
> The Rust book/site has always felt a bit old to me.

Yeah agreed, the Rust book is built with `mdbook` which is written in Rust and has a really nice dev experience, but it's missing all the modern web niceties so not great for user experience. My favorite [doc site is for Nushell](https://www.nushell.sh/book/) which is built with Vuepress 2 (currently in beta) and also what I use for [mojodojo.dev](https://mojodojo.dev). Can see the [source code here](https://github.com/mojodojodev/mojodojo.dev) if anyone wants to copy paste that for their own book or docs, I've kept it really simple and put what everything does in the README.md.

**Comment by ksandvik** [comment]
Docusaurus is also a very neat documentation package where you could something nice up and running in minutes. The key is to always keep the documentation in .md format so it could be re-used by various tools or in case you want to switch to another static web generator or publishing tool.
